sql: default optimizer to "on"

justinj opened the pull request:
This commit sets the cluster setting `sql.defaults.optimizer` to `'on'`
by default.

We now defer any queries involving the JSONB operators to the heuristic
planner in order to make sure inverted index queries still work. A bunch
of tests that use them have been commented out until we can either plan
the queries with the optimizer or have a less brute-force method to fall
back here. For now I just did the simplest possible thing. This was necessary
to make some tests pass.

A number of tests also needed to be updated to work with the optimizer.

It's possible there's still some lurking problems here, but hopefully by
rolling this out internally we can find them.

If we decide to turn this off before the next alpha we can still do
that.

Release note (sql change): The cost-based optimizer is now enabled by
default. It can be turned off by setting the `sql.defaults.optimizer`
cluster setting to `'off'`.

Commit:
opt: fall back for queries with json ops

This commit causes queries which involve JSON operators to fall back to
the heuristic planner. This involved jumping through a few hoops since
the heuristic planner uses the opt span generation code for inverted
indexes.

Release note: None.

Commit:
sql: modify tests to pass with optimizer on

This commit changes some tests that were depending on the plans
generated by the existing planner to still pass when the cost-based
optimizer is enabled.

Release note: None.

Commit:
sql: turn optimizer on by default

This commit sets the cluster setting `sql.defaults.optimizer` to `on` by
default.

Tests still pass with it off, so we can roll this back if need be.

Also remove the "always" mode.

Release note (sql change): The cost-based optimizer is now on by
default.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26893)
<!-- Reviewable:end -->

knz commented:
Very nice patch. LGTM with nits.

---

Reviewed 13 of 13 files at r1, 6 of 6 files at r2, 1 of 1 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/cli/cli_test.go, line 1624 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfMTg1XYb9oLuAYcO:-LFXfMTg1XYb9oLuAYcP:b-bbqsv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca0e2b60f26b67cee7550f913a2d2e0f7cc2814/pkg/cli/cli_test.go#L1624)):*
> ```Go
> 	// (1 row)
> 	// sql --format=pretty -e explain select s from t.t union all select s from t.t
> 	// +----------------+-------+-------------+
> ```

Do not remove this test, or replace it with an equivalent test: it is meant to verify that the special Unicode tree characters pass through the entire SQL client stack to the terminal.

---

*[pkg/sql/exec_util.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfXZ03zWqUkenASIc:-LFXfXZ03zWqUkenASId:bg1n6ya) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/exec_util.go#L108)):*
> ```Go
> 	"sql.defaults.optimizer",
> 	"Default cost-based optimizer mode",
> 	"On",
> ```

Here and below - the values should be lowercased.

---

*[pkg/sql/trace_test.go, line 167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfeYQ6gfWcY--Z_Bj:-LFXfeYQ6gfWcY--Z_Bk:b-2bs2lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L167)):*
> ```Go
> 					t.Fatal(err)
> 				}
> 				// TODO(justin): remove this and make sure the new results make sense.
> ```

Can you detail what happens if you do not disable the optimizer here? Based on the code below I should expect there is no difference.

---

*[pkg/sql/logictest/logic.go, line 388 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfqtsAVLjOXYqXEiB:-LFXfqtsAVLjOXYqXEiC:bhu6vag) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/logictest/logic.go#L388)):*
> ```Go
> // via -config).
> var logicTestConfigs = []testClusterConfig{
> 	{name: "local", numNodes: 1, overrideDistSQLMode: "Off", overrideOptimizerMode: "Off"},
> ```

Here and below the constants should be lowercased.

---

*[pkg/sql/opt/memo/statistics_builder.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXg20T8hPZBlVvKejQ:-LFXg20T8hPZBlVvKejR:b48zzqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/opt/memo/statistics_builder.go#L164)):*
> ```Go
> 		return sb.colStatRowNumber(colSet)
> 
> 	case opt.ExplainOp, opt.ShowTraceOp, opt.ShowTraceForSessionOp:
> ```

Why did you need to change this? Perhaps it belong to a separate commit with a dedicated explanatory commit message?

---

*[pkg/sql/pgwire/pgwire_test.go, line 822 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXgMZN3pcYi-C2UfdB:-LFXgMZN3pcYi-C2UfdC:b-yg8702) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/pgwire/pgwire_test.go#L822)):*
> ```Go
> 			baseTest.SetArgs(45).Results(45),
> 		},
> 		// TODO(justin): match this with the optimizer. Currently we only report
> ```

Perhaps you want to file an issue to track this.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LFXgP5H4U5gcCkfI4bP:b-qdny3a)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very exciting! 

I'm not sure that the `AllowBlacklistOps` flag is necessary

---

Reviewed 13 of 13 files at r1, 6 of 6 files at r2, 1 of 1 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/optbuilder/scalar.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXpxE78R7Xlsl-17_Y:-LFXpxE85uVP-r_swG9e:bk1qp60) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a92fbcc2e00358697f6b767bfe7b6a754aa4721/pkg/sql/opt/optbuilder/scalar.go#L69)):*
> ```Go
> // planner uses the same constraint generation code.
> // Not all of these generate inverted index scans today.
> var comparisonOpBlacklist = [tree.NumComparisonOperators]bool{
> ```

I think you can just remove them from the above map, because then it would rely on the "AllowUnsupported" flag

---

*[pkg/sql/opt/optbuilder/scalar.go, line 102 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXqF77Da4xwkv3hnZU:-LFXqF77Da4xwkv3hnZV:b-89a4l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a92fbcc2e00358697f6b767bfe7b6a754aa4721/pkg/sql/opt/optbuilder/scalar.go#L102)):*
> ```Go
> // planner uses the same constraint generation code.
> // Not all of these generate inverted index scans today.
> var binaryOpBlacklist = [tree.NumBinaryOperators]bool{
> ```

Ditto

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LFXrSDN7oRtod2qw4wR:b-e7bbfx)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Nice work weeding out all the issues!

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/sql/trace_test.go, line 167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfeYQ6gfWcY--Z_Bj:-LFXu0LG34esRhejq3Zl:bavmjia) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L167)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you detail what happens if you do not disable the optimizer here? Based on the code below I should expect there is no difference.
</blockquote></details>

I'm guessing it was crashing because of the ShowTraceOp stat issue, it should work now?

---

*[pkg/sql/trace_test.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXuIGmDGhgU89fn4Rm:-LFXuIGmDGhgU89fn4Rn:b-z4on73) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L172)):*
> ```Go
> 				}
> 				return sqlDB.Query(
> 					"SELECT DISTINCT(operation) op FROM [SHOW TRACE FOR SELECT * FROM test.foo] " +
> ```

[nit] this query is very badly formatted. DISTINCT is not a function. It should be `SELECT DISTINCT operation AS op FROM`

---

*[pkg/sql/opt/memo/statistics_builder.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXg20T8hPZBlVvKejQ:-LFXtddz27sdDZSDB3Ex:bkoxsul) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/opt/memo/statistics_builder.go#L164)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why did you need to change this? Perhaps it belong to a separate commit with a dedicated explanatory commit message?
</blockquote></details>

This should have been in #26749. Your other PR is removing the op but it hasn't merged yet..

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LFXu_srDh6pnA5pfS5h:b-te9zo5)*
<!-- Sent from Reviewable.io -->

justinj commented:
Addressed comments, turns out still some acceptance tests I need to fix.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/cli/cli_test.go, line 1624 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfMTg1XYb9oLuAYcO:-LFXs7k_6DW7HVZ9M34N:b-3xg90a) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ca0e2b60f26b67cee7550f913a2d2e0f7cc2814/pkg/cli/cli_test.go#L1624)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Do not remove this test, or replace it with an equivalent test: it is meant to verify that the special Unicode tree characters pass through the entire SQL client stack to the terminal.
</blockquote></details>

Ok, I changed it to a test which still has tree characters, but has the same plan across both planners.

---

*[pkg/sql/exec_util.go, line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfXZ03zWqUkenASIc:-LFXtNE519RzAZHLqEEo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/exec_util.go#L108)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here and below - the values should be lowercased.
</blockquote></details>

Done.

---

*[pkg/sql/trace_test.go, line 167 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfeYQ6gfWcY--Z_Bj:-LFXtQhT2U-Go5OPZSH3:byw89xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L167)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you detail what happens if you do not disable the optimizer here? Based on the code below I should expect there is no difference.
</blockquote></details>

The optimizer elides some renders that the heuristic planner doesn't which makes the output here different. I added a comment to that effect.

---

*[pkg/sql/trace_test.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXuIGmDGhgU89fn4Rm:-LFYjwv3DyUb0vV4W2Iq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L172)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this query is very badly formatted. DISTINCT is not a function. It should be `SELECT DISTINCT operation AS op FROM`
</blockquote></details>

Done.

---

*[pkg/sql/logictest/logic.go, line 388 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXfqtsAVLjOXYqXEiB:-LFXu0NcCcPVsoWqsTTa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/logictest/logic.go#L388)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Here and below the constants should be lowercased.
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/statistics_builder.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXg20T8hPZBlVvKejQ:-LFXvJPW0dcNMj5IGvEi:b-3w1n0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/opt/memo/statistics_builder.go#L164)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should have been in #26749. Your other PR is removing the op but it hasn't merged yet..
</blockquote></details>

Is `ShowTraceOp` going away? Which PR should I wait on for that?

---

*[pkg/sql/opt/optbuilder/scalar.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXpxE78R7Xlsl-17_Y:-LFXwZUF0cK9MDeJxpXv:b-cy59c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a92fbcc2e00358697f6b767bfe7b6a754aa4721/pkg/sql/opt/optbuilder/scalar.go#L69)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think you can just remove them from the above map, because then it would rely on the "AllowUnsupported" flag
</blockquote></details>

The `UnsupportedExpr` op that gets generated in that case doesn't get treated properly for span generation, as far as I can tell, we need to actually generate the `ContainsOp` in that case. Unless I misunderstand what you mean?

---

*[pkg/sql/pgwire/pgwire_test.go, line 822 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXgMZN3pcYi-C2UfdB:-LFYiHYm4nuuQUnmVUX9:btos4f3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/pgwire/pgwire_test.go#L822)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Perhaps you want to file an issue to track this.
</blockquote></details>

opened as #26901.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LFYlS239QBssoxhlgZt:b-eonqaz)*
<!-- Sent from Reviewable.io -->

rytaft commented:
*[pkg/sql/opt/optbuilder/scalar.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXpxE78R7Xlsl-17_Y:-LFYuWpf7IB0fSaIBB84:bakmi38) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a92fbcc2e00358697f6b767bfe7b6a754aa4721/pkg/sql/opt/optbuilder/scalar.go#L69)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

The `UnsupportedExpr` op that gets generated in that case doesn't get treated properly for span generation, as far as I can tell, we need to actually generate the `ContainsOp` in that case. Unless I misunderstand what you mean?
</blockquote></details>

I'm not 100% sure, so you're probably right. Thanks!

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/memo/statistics_builder.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXg20T8hPZBlVvKejQ:-LFYvyh1-GeATofh8V-I:b-bkma1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/opt/memo/statistics_builder.go#L164)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Is `ShowTraceOp` going away? Which PR should I wait on for that?
</blockquote></details>

#26729 but not sure how close it is to landing. I'd move forward with this change.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks! 

---

Reviewed 4 of 5 files at r4, 2 of 2 files at r5.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/trace_test.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXuIGmDGhgU89fn4Rm:-LF_NyEqBJSWz_Yjbc_G:bxberiz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L172)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see any change here.

---

*[pkg/sql/opt/memo/statistics_builder.go, line 164 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXg20T8hPZBlVvKejQ:-LF_O3ME1_FjxEtxr_0H:bnwt3pg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/opt/memo/statistics_builder.go#L164)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

#26729 but not sure how close it is to landing. I'd move forward with this change.
</blockquote></details>

Don't wait. I'll rebase.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LF_O9sN5ayaRByfKiFg:b-hay2ib)*
<!-- Sent from Reviewable.io -->

Commit:
opt: pass through audit events and disable test

We need to pass on the list of audit events generated during planning
for them to materialize in the logs. This commit makes this change.

This commit also disables an audit logging test involving logging an
error when selecting from an unknown column. Fixing this is tracked in
issue #27036.

Release note: None

knz commented:
Thanks for letting me know. I presume you are advancing on this PR fast and hard so as to get an auto-on optimizer in the July alpha. Is that correct? 

If so, I must push back because I am of the opinion there is no way we should publish an alpha with an auto-on feature that disables privilege checks altogether. Unless you also fix #27036 in the same PR. @bdarnell what do you think?

---

Reviewed 2 of 9 files at r6, 5 of 6 files at r7, 3 of 3 files at r8, 2 of 2 files at r9.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/trace_test.go, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-LFXuIGmDGhgU89fn4Rm:-LG4riAe1Jowt4h7dT49:bcbxlqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2bf79fdffc1c2c5a16856c87ef73d82d80d677f4/pkg/sql/trace_test.go#L172)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't see any change here.
</blockquote></details>

`SELECT DISTINCT operation AS op FROM ...`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LG4s38mAZlt1Xifyca4:byzewpo)*
<!-- Sent from Reviewable.io -->

knz commented:
Oh nevermind. Peter just taught me the release SHA was selected already. So my concerns are alleviated.

<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>  thanks!

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LG515F614Um2260oIB3:b-k2gye9)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Correct, we didn't make this alpha but we want to turn it on as soon as possible to weed out any major issues before the next alpha.

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26893#-:-LG5OYTVBDOwWie9Le4Y:bxs9vto)*
<!-- Sent from Reviewable.io -->

justinj commented:
Cool, thanks! Definitely agree #27036 is a non-starter for a release.

Fixed the distinct thing - there were actually a couple instances in that test and I had only changed one.

justinj commented:
bors r+

justinj commented:
Really hoping this is the last try -

bors r+
